config/src/main/java/com/typesafe/config/impl/ConfigParser.java

```diff
@@ -0,0 +1,435 @@
+/**
+ *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
+ */
+package com.typesafe.config.impl;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Stack;
+
+import com.typesafe.config.*;
+
+final class ConfigParser {
+    static AbstractConfigValue parse(ConfigNodeRoot document,
+                                     ConfigOrigin origin, ConfigParseOptions options,
+                                     ConfigIncludeContext includeContext) {
+        ParseContext context = new ParseContext(options.getSyntax(), origin, document,
+                SimpleIncluder.makeFull(options.getIncluder()), includeContext);
+        return context.parse();
+    }
+
+    static private final class ParseContext {
+        private int lineNumber;
+        final private ConfigNodeRoot document;
+        final private FullIncluder includer;
+        final private ConfigIncludeContext includeContext;
+        final private ConfigSyntax flavor;
+        final private ConfigOrigin baseOrigin;
```

Comment:
I wonder if we can track lineNumber and origin on each node, and flavor on the document, so those three can be omitted here ?

IIRC some of the tokens (such as braces) don't track line and origin in an attempt to be more efficient; but we can maybe compute line and origin by starting from a token that does have them and then scanning looking at newlines, or something.
I'm not sure exactly how to do it. I guess we could leave this as a future enhancement.